File: IAIso-v5.0/core/iaiso-php/tests/Unit/Policy/PolicyTest.php
<?php

declare(strict_types=1);

namespace IAIso\Tests\Unit\Policy;

use IAIso\Policy\MaxAggregator;
use IAIso\Policy\PolicyException;
use IAIso\Policy\PolicyLoader;
use IAIso\Policy\SumAggregator;
use IAIso\Policy\WeightedSumAggregator;
use PHPUnit\Framework\TestCase;

final class PolicyTest extends TestCase
{
    public function testBuildMinimalPolicy(): void
    {
        $p = PolicyLoader::build(['version' => '1']);
        self::assertSame('1', $p->version);
        self::assertSame('sum', $p->aggregator->name());
    }

    public function testBuildOverridesDefaults(): void
    {
        $p = PolicyLoader::build([
            'version' => '1',
            'pressure' => ['escalation_threshold' => 0.7, 'release_threshold' => 0.85],
            'coordinator' => ['aggregator' => 'max'],
        ]);
        self::assertEqualsWithDelta(0.7, $p->pressure->escalationThreshold, 1e-9);
        self::assertSame('max', $p->aggregator->name());
    }

    public function testRejectsMissingVersion(): void
    {
        $this->expectException(PolicyException::class);
        $this->expectExceptionMessageMatches('/version/');
        // Use stdClass-like associative array so it's seen as a mapping
        // but with no `version` key.
        PolicyLoader::build(['metadata' => ['note' => 'no version here']]);
    }

    public function testRejectsBadVersion(): void
    {
        $this->expectException(PolicyException::class);
        PolicyLoader::build(['version' => '2']);
    }

    public function testRejectsReleaseBelowEscalation(): void
    {
        $this->expectException(PolicyException::class);
        PolicyLoader::build([
            'version' => '1',
            'pressure' => ['escalation_threshold' => 0.9, 'release_threshold' => 0.5],
        ]);
    }

    public function testRejectsStringAsNumber(): void
    {
        // The strict-typing gotcha: string "0.015" should NOT validate as a number.
        $this->expectException(PolicyException::class);
        PolicyLoader::build([
            'version' => '1',
            'pressure' => ['token_coefficient' => '0.015'],
        ]);
    }

    public function testSumAggregator(): void
    {
        self::assertEqualsWithDelta(
            0.8, (new SumAggregator())->aggregate(['a' => 0.3, 'b' => 0.5]), 1e-9);
    }

    public function testMaxAggregator(): void
    {
        self::assertEqualsWithDelta(
            0.5, (new MaxAggregator())->aggregate(['a' => 0.3, 'b' => 0.5]), 1e-9);
    }

    public function testWeightedSumAggregator(): void
    {
        $a = new WeightedSumAggregator(['important' => 2.0], 1.0);
        self::assertEqualsWithDelta(
            1.3, $a->aggregate(['important' => 0.5, 'normal' => 0.3]), 1e-9);
    }
}

File: IAIso-v5.0/core/iaiso-ruby/test/policy_test.rb
# frozen_string_literal: true

require_relative "test_helper"
require "iaiso/policy"

class PolicyTest < Minitest::Test
  def test_build_minimal_policy
    p = IAIso::Policy::Loader.build({ "version" => "1" })
    assert_equal "1", p.version
    assert_equal "sum", p.aggregator.name
  end

  def test_build_overrides_defaults
    p = IAIso::Policy::Loader.build({
      "version" => "1",
      "pressure" => { "escalation_threshold" => 0.7, "release_threshold" => 0.85 },
      "coordinator" => { "aggregator" => "max" },
    })
    assert_in_delta 0.7, p.pressure.escalation_threshold, 1e-9
    assert_equal "max", p.aggregator.name
  end

  def test_rejects_missing_version
    err = assert_raises(IAIso::Policy::PolicyError) do
      IAIso::Policy::Loader.build({ "metadata" => { "note" => "no version" } })
    end
    assert_match(/version/, err.message)
  end

  def test_rejects_bad_version
    assert_raises(IAIso::Policy::PolicyError) do
      IAIso::Policy::Loader.build({ "version" => "2" })
    end
  end

  def test_rejects_release_below_escalation
    assert_raises(IAIso::Policy::PolicyError) do
      IAIso::Policy::Loader.build({
        "version" => "1",
        "pressure" => { "escalation_threshold" => 0.9, "release_threshold" => 0.5 },
      })
    end
  end

  def test_rejects_string_as_number
    # Strict typing: string "0.015" should NOT validate as a number.
    assert_raises(IAIso::Policy::PolicyError) do
      IAIso::Policy::Loader.build({
        "version" => "1",
        "pressure" => { "token_coefficient" => "0.015" },
      })
    end
  end

  def test_rejects_top_level_array
    assert_raises(IAIso::Policy::PolicyError) do
      IAIso::Policy::Loader.build([1, 2, 3])
    end
  end

  def test_sum_aggregator
    assert_in_delta 0.8, IAIso::Policy::SumAggregator.new.aggregate({ "a" => 0.3, "b" => 0.5 }), 1e-9
  end

  def test_max_aggregator
    assert_in_delta 0.5, IAIso::Policy::MaxAggregator.new.aggregate({ "a" => 0.3, "b" => 0.5 }), 1e-9
  end

  def test_weighted_sum_aggregator
    a = IAIso::Policy::WeightedSumAggregator.new(weights: { "important" => 2.0 }, default_weight: 1.0)
    assert_in_delta 1.3, a.aggregate({ "important" => 0.5, "normal" => 0.3 }), 1e-9
  end
end
